File: backend/src/main/java/com/jhub/backend/model/JobApplication.java
package com.jhub.backend.model;

import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import com.jhub.backend.model.enums.JobApplicationStatus;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

/**
 * represents a job application tracked by a user
 * each user can have many job applications, and each job application can have many notes
 */
@Entity
@Table(name = "job_applications")
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@ToString(exclude = { "user", "notes" })
@EqualsAndHashCode(of = "id")
public class JobApplication {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 255)
    @NotBlank(message = "Company name is required")
    @Size(max = 255, message = "Company name must not exceed 255 characters")
    private String company;

    @Column(nullable = false, length = 255)
    @NotBlank(message = "Job title is required")
    @Size(max = 255, message = "Job title must not exceed 255 characters")
    private String jobTitle;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private JobApplicationStatus status = JobApplicationStatus.APPLIED;

    @Column(nullable = false)
    private LocalDate dateApplied;

    @Column(length = 2048)
    @Size(max = 2048, message = "Job posting URL must not exceed 2048 characters")
    private String jobPostingUrl;

    @Column(length = 255)
    @Size(max = 255, message = "Location must not exceed 255 characters")
    private String location;

    @Min(value = 0, message = "Minimum salary must not be negative")
    private Integer salaryMin;

    @Min(value = 0, message = "Maximum salary must not be negative")
    private Integer salaryMax;

    @Column(columnDefinition = "TEXT")
    private String description;

    @OneToMany(
        mappedBy = "jobApplication",
        cascade = CascadeType.ALL,
        orphanRemoval = true
    )
    private List<Note> notes = new ArrayList<>();

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private Instant createdAt;

    @LastModifiedDate
    @Column(nullable = false)
    private Instant updatedAt;

    @Builder
    public JobApplication(
        User user,
        String company,
        String jobTitle,
        JobApplicationStatus status,
        LocalDate dateApplied,
        String jobPostingUrl,
        String location,
        Integer salaryMin,
        Integer salaryMax,
        String description
    ) {
        this.user = user;
        this.company = company;
        this.jobTitle = jobTitle;
        this.status = status != null ? status : JobApplicationStatus.APPLIED;
        this.dateApplied = dateApplied;
        this.jobPostingUrl = jobPostingUrl;
        this.location = location;
        this.salaryMin = salaryMin;
        this.salaryMax = salaryMax;
        this.description = description;
    }

    public void addNote(Note note) {
        notes.add(note);
        note.setJobApplication(this);
    }

    public void removeNote(Note note) {
        notes.remove(note);
        note.setJobApplication(null);
    }
}

File: frontend/src/lib/schemas/jobApplication.ts
import { z } from 'zod/v4';
import { JOB_APPLICATION_STATUSES } from '@/lib/types';

export const jobApplicationSchema = z
  .object({
    company: z
      .string()
      .min(1, 'Company name is required')
      .max(255, 'Company name must be 255 characters or less'),
    jobTitle: z
      .string()
      .min(1, 'Job title is required')
      .max(255, 'Job title must be 255 characters or less'),
    dateApplied: z.date({ error: 'Application date is required' }),
    status: z.enum(JOB_APPLICATION_STATUSES),
    jobPostingUrl: z
      .string()
      .max(2048, 'URL must be 2048 characters or less')
      .refine((val) => val === '' || z.string().url().safeParse(val).success, 'Enter a valid URL'),
    location: z.string().max(255, 'Location must be 255 characters or less'),
    salaryMin: z.number().nonnegative('Salary must be a positive number').nullable(),
    salaryMax: z.number().nonnegative('Salary must be a positive number').nullable(),
    description: z.string(),
  })
  .refine(
    (data) => {
      if (data.salaryMin != null && data.salaryMax != null) {
        return data.salaryMin <= data.salaryMax;
      }
      return true;
    },
    { message: 'Minimum salary must not exceed maximum salary', path: ['salaryMin'] }
  );

export type JobApplicationFormValues = z.infer<typeof jobApplicationSchema>;
